### PR TITLE
Fix #4014 - Cannot download Components from remote BCL: handle crash when times out + add ability to change timeout on http client

### LINKED ISSUE #4014: Can't download components from BCL (using OpenStudio Application 1.0)
Issue overview
--------------
Reported by user here
https://unmethours.com/question/45690/cannot-download-components-from-bcl-in-openstudio-application-if-downloaded-direct-from-website-how-can-i-import-them-into-the-application/

I was able to reproduce on my computer for same material they mentioned, as well as a few others.

## Current Behavior
There is another issue where search doesn't work for some users. I was able to search for brick and select the material and click download. The download progress bar doesn't stop but I thought that was just GUI glitch. The component is shown as already downloaded with grey checkmark but isn't in BCL directory. User can't try to download again since it shows as already having it. I did same with a few more materials, so isn't unique to this.

view of app after attempt to download material, not progress bar doesn't stop moving, but checkmark changes form black to grey (which typically indicates download is complete)
Screen Shot 2020-07-08 at 9 40 56 AM

View of app and file manager after 3 materials. Note the one object in directory is OpenStudio Results measure.
Screen Shot 2020-07-08 at 9 43 17 AM

## Expected Behavior
Download to BCL directory and show in GUI at end of "Library" for materials

## Steps to Reproduce

Try to download using OpenStudio Application

## Possible Solution

Work around is to download with web browser (which other user was able to do) and then to load OSM as library.

### Environment

Some additional details about your environment for this issue (if relevant):
 - Platform (Operating system, version): Mac 10.15.3
 - Version of OpenStudio (if using an intermediate build, include SHA): OpenStudio Application 1.0

### PR DESCRIPTION
**Pull request by jmarrec**
Pull request overview
---------------------

- Fix #4014 - Cannot download Components frrom remote BCL

#4021 had increased timeout from 30 to 60secs, but BCL is super slow **for components** (not measures) so increasing it to 120secs. Also cleaning up the RemoteBCL class and its use of cpprestsdk to avoid a crash and allow changing the timeout manually

### Pull Request Author

Add to this list or remove from it as applicable.  This is a simple templated set of guidelines.

 - [ ] Model API Changes / Additions
 - [ ] Any new or modified fields have been implemented in the EnergyPlus ForwardTranslator (and ReverseTranslator as appropriate)
 - [ ] Model API methods are tested (in `src/model/test`)
 - [ ] EnergyPlus ForwardTranslator Tests (in `src/energyplus/Test`)
 - [ ] If a new object or method, added a test in NREL/OpenStudio-resources: Add Link
 - [ ] If needed, added VersionTranslation rules for the objects (`src/osversion/VersionTranslator.cpp`)
 - [ ] Checked behavior in OpenStudioApplication, adjusted policies as needed (`src/openstudio_lib/library/OpenStudioPolicy.xml`)
 - [ ] Verified that C# bindings built fine on Windows, partial classes used as needed, etc.
 - [ ] All new and existing tests passes
 - [ ] If methods have been deprecated, update rest of code to use the new methods

**Labels:**

 - [ ] If change to an IDD file, add the label `IDDChange`
 - [ ] If breaking existing API, add the label `APIChange`
 - [ ] If deemed ready, add label `Pull Request - Ready for CI` so that CI builds your PR

### Review Checklist

This will not be exhaustively relevant to every PR.
 - [ ] Perform a Code Review on GitHub
 - [ ] Code Style, strip trailing whitespace, etc.
 - [ ] All related changes have been implemented: model changes, model tests, FT changes, FT tests, VersionTranslation, OS App
 - [ ] Labeling is ok
 - [ ] If defect, verify by running develop branch and reproducing defect, then running PR and reproducing fix
 - [ ] If feature, test running new feature, try creative ways to break it
 - [ ] CI status: all green or justified

### CODE REVIEW

File: src/utilities/bcl/test/BCL_GTest.cpp [hunk 1]
```diff
@@ -429,3 +429,36 @@ TEST_F(BCLFixture, RemoteBCLMetaSearchTest) {
...
+  EXPECT_EQ(11u, remoteBCL.timeOutSeconds());
+
+  bool success = remoteBCL.downloadComponent(uid);
+  ASSERT_TRUE(success);
+
+  boost::optional<BCLComponent> component = remoteBCL.waitForComponentDownload();
+  EXPECT_FALSE(component);
+
+  // Try again, should not segfault
+  success = remoteBCL.downloadComponent(uid);
```
Comment (jmarrec) [inline src/utilities/bcl/test/BCL_GTest.cpp:463]: Before: it crashes

```
Trace/breakpoint trap (core dumped)
```

File: src/utilities/bcl/RemoteBCL.cpp [hunk 2]
```diff
@@ -98,7 +112,8 @@ RemoteBCL::RemoteBCL()
     m_lastTotalResults(0),
     m_apiVersion("2.0"),
     validProdAuthKey(false),
-    validDevAuthKey(false) {
+    validDevAuthKey(false),
+    m_timeOutSeconds(120) {
```
Comment (jmarrec) [inline src/utilities/bcl/RemoteBCL.cpp:116]: Actual fix for #4014: currently downloading Components is super slow, takes 66 seconds on my machine. Downloading Measures on the other hand is much faster...

File: src/utilities/bcl/RemoteBCL.cpp [hunk 3]
```diff
@@ -763,36 +798,22 @@ bool RemoteBCL::startComponentLibrarySearch(const std::string& searchTerm, const
...
+  if (!m_httpResponse) {
+    return false;
+  }
 
-    ++current;
+  try {
+    m_httpResponse->wait();
+    return true;
+  } catch (const std::exception& e) {
+    LOG(Error, "Request to url '" << m_remoteUrl << " 'failed with message: " << e.what());
```
Comment (jmarrec) [inline src/utilities/bcl/RemoteBCL.cpp:816]: Fix the crash portion I found while investigating #4014 and trying twice to download

File: src/utilities/bcl/RemoteBCL.hpp [hunk 4]
```diff
@@ -178,21 +179,28 @@ class UTILITIES_API RemoteBCL : public BCL
...
   int numResultPages() const;
 
+  unsigned timeOutSeconds() const;
+  bool setTimeOutSeconds(unsigned timeOutSeconds);
+
   /// Wait number of milliseconds for download to complete
   /// Returns the download if it completed in the allowable time
-  boost::optional<BCLComponent> waitForComponentDownload(int msec = 120000) const;
+  boost::optional<BCLComponent> waitForComponentDownload() const;
+  OS_DEPRECATED boost::optional<BCLComponent> waitForComponentDownload(int) const;
```
Comment (jmarrec) [inline src/utilities/bcl/RemoteBCL.hpp:188]: Deprecated the msec version of these

File: src/utilities/bcl/RemoteBCL.hpp [hunk 5]
```diff
@@ -178,21 +179,28 @@ class UTILITIES_API RemoteBCL : public BCL
   /// Return the number of pages of results
   int numResultPages() const;
 
+  unsigned timeOutSeconds() const;
+  bool setTimeOutSeconds(unsigned timeOutSeconds);
```
Comment (jmarrec) [inline src/utilities/bcl/RemoteBCL.hpp:183]: Add getter/setter to change RemoteBCL timeout

File: src/utilities/bcl/RemoteBCL.hpp [hunk 6]
```diff
@@ -250,7 +258,7 @@ class UTILITIES_API RemoteBCL : public BCL
   // members
 
   // A helper function to prepare a client, allowing us to change the http_client_config in one place only
-  static web::http::client::http_client getClient(const std::string& url);
+  static web::http::client::http_client getClient(const std::string& url, unsigned timeOutSeconds = 60);
```
Comment (jmarrec) [inline src/utilities/bcl/RemoteBCL.hpp:261]: static getClient  now also takes an optional timeOutSeconds. This one is set to 60 to match prior behavior, even thought I icnreased the RemoteBCL timeout to 120

File: src/utilities/bcl/RemoteBCL.cpp [hunk 7]
```diff
@@ -693,7 +728,7 @@ bool RemoteBCL::startComponentLibrarySearch(const std::string& searchTerm, const
 
   m_lastSearch.clear();
 
-  auto client = getClient(remoteUrl());
+  auto client = getClient(remoteUrl(), m_timeOutSeconds);
```
Comment (jmarrec) [inline src/utilities/bcl/RemoteBCL.cpp:731]: Every internal call to getClient in RemoteBCL passes the member m_timeOutSeconds now

File: src/utilities/bcl/test/BCL_GTest.cpp (hunk 1 above)
Comment (jmarrec) [reply src/utilities/bcl/test/BCL_GTest.cpp:463]: After: passes

Log messages

```
[ RUN      ] BCLFixture.4014_Crash
[openstudio.RemoteBCL] <0> Setting a timeout of 11 appears low and you risk failures to download components and measures
[openstudio.RemoteBCL] <1> Request to url 'https://bcl.nrel.gov 'failed with message: Failed to read HTTP status line
[       OK ] BCLFixture.4014_Crash (12006 ms)

```

**Review by kbenne — APPROVED** [review]

**Comment by jmarrec** [comment]
@kbenne Should we merge?

**Comment by kbenne** [comment]
@jmarrec I'm going to leave that up to @tijcolem but the changes look good to me.

**Comment by tijcolem** [comment]
Looks good to me!  This will be nice to have.

**Pull request merged**